File: app/src/main/java/ru/arhlib/app/actual/ActualFragment.java
package ru.arhlib.app.actual;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import ru.arhlib.app.R;
import ru.arhlib.app.databinding.ActualFragmentBinding;

public class ActualFragment extends Fragment {

    public ActualFragment() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ActualFragmentBinding binding = ActualFragmentBinding.inflate(inflater, container, false);
        View view = binding.getRoot();

        ActualViewModel viewModel = ViewModelProviders.of(this).get(ActualViewModel.class);
        RecyclerView recyclerView = view.findViewById(R.id.actual_list);
        ActualAdapter adapter = new ActualAdapter();
        viewModel.actual.observe(this, adapter::submitList);
        viewModel.showErrorSnackbar.observe(this, (needToShow) -> {
            if (needToShow) {
                Snackbar.make(view, R.string.loading_error, Snackbar.LENGTH_SHORT).show();
                viewModel.showErrorSnackbar.postValue(false);
            }
        });
        recyclerView.setAdapter(adapter);
        binding.setViewModel(viewModel);

        return view;
    }
}

File: app/src/main/java/ru/arhlib/app/actual/ActualFragment.kt
package ru.arhlib.app.actual

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import ru.arhlib.app.databinding.ActualFragmentBinding

class ActualFragment : Fragment() {
    private val viewModel: ActualViewModel by viewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val binding = ActualFragmentBinding.inflate(inflater, container, false)
        val adapter = ActualItemAdapter()
        viewModel.actualItems.observe(viewLifecycleOwner, { result ->
            if (result is LoadResult.Success) {
                adapter.submitList(result.data)
                binding.emptyActualText.isVisible = result.data.isEmpty()
                binding.actualList.isVisible = result.data.isNotEmpty()
            }
            if (result is LoadResult.Error) {
                binding.actualList.isVisible = false
            }
            binding.progressBar.isVisible = result is LoadResult.Loading
            binding.retryBlock.isVisible = result is LoadResult.Error
            binding.swipeRefresh.isRefreshing = result is LoadResult.Refreshing
        })
        binding.actualList.adapter = adapter
        binding.retryButton.setOnClickListener {
            viewModel.load()
        }
        binding.swipeRefresh.setOnRefreshListener {
            viewModel.refresh()
        }
        return binding.root
    }
}
